### PR TITLE
Model and ViewModel for message-list feature.

### PR DESCRIPTION
**Pull request by BarAgam**
Made the ViewModel and the Model parts of the messages management use-case (for messages displaying, choosing selected message, and read status of different messages).
MessagesViewModel needs MessagesModel. MessagesReadStatusModel and SelectedMessageModel to function.
The models have been implemented with an in-memory implementation. The app will use these implementations, until a better alternative (DataBase for exmaple) will be implemented. The tests use the in-memory implementations, even if another implementation will be created, since they're perfect for testing.

Both the InMemoryMessageModel and MessagesViewModel class have test-classes. The other classes are mostly data objects/ simple enough not to be tested.

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/teamagam/GimelGimel/pull/56" target="_blank">ReviewNinja

File: app/src/main/java/com/teamagam/gimelgimel/app/common/DataChangedObservable.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.teamagam.gimelgimel.app.common;
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/common/DataChangedObservable.java:1]: dont you think that this should be in model dir?

File: app/src/main/java/com/teamagam/gimelgimel/app/common/DataChangedObserver.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.teamagam.gimelgimel.app.common;
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/common/DataChangedObserver.java:1]: same

File: app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package com.teamagam.gimelgimel.app.common;
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java:1]: ...

File: app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * A simple implementation of an observable, that supports a single observer
+ * and a notifying method
+ */
+public class NotifyingDataChangedObservable implements DataChangedObservable {
+
+    private DataChangedObserver mObserver;
+
+    @Override
+    public void setObserver(DataChangedObserver observer) {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java]: where do you manage list of observers?

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/Message.java [hunk 5]
```diff
@@ -43,40 +43,19 @@ public T getContent() {
         return mContent;
     }
 
-    public void setContent(T content) {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/Message.java:46]: why is that?? we still need some of these functionality.

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/DisplayMessage.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+import com.teamagam.gimelgimel.app.model.ViewsModels.Message;
+
+/**
+ * Data class for message displaying
+ */
+public class DisplayMessage {
+
+    private Message mMessage;
+    private boolean mIsSelected;
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/DisplayMessage.java:11]: HaHa mIsRead for each message ;)

Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/DisplayMessage.java:11]: I mean mIsSelected

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/Message.java [hunk 7]
```diff
@@ -43,40 +43,19 @@ public T getContent() {
...
 
-    public void setSenderId(String senderId) {
-        mSenderId = senderId;
-    }
-
     public Date getCreatedAt() {
         return mCreatedAt;
     }
 
-    public void setCreatedAt(Date createdAt) {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/Message.java:70]: for example

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+import com.teamagam.gimelgimel.app.common.NotifyingDataChangedObservable;
+import com.teamagam.gimelgimel.app.model.ViewsModels.Message;
+import com.teamagam.gimelgimel.app.model.entities.messages.MessagesModel;
+import com.teamagam.gimelgimel.app.model.entities.messages.MessagesReadStatusModel;
+import com.teamagam.gimelgimel.app.model.entities.messages.SelectedMessageModel;
+
+/**
+ * Messages view-model for messages presentation use-case
+ */
+public class MessagesViewModel extends NotifyingDataChangedObservable implements DataChangedObservable {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java:15]: I dont like that this VM extends NotifyingDataChangedObservable.

consider changing It's name to BaseViewModel !resolve

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                return mMessagesReadStatusModel.isRead(message);
+            }
+
+            private boolean isSelected(Message message, int index) {
+                if (mSelectedMessageModel.isAnySelected()) {
+                    return mSelectedMessageModel.getSelected() == message;
+                }
+
+                mSelectedMessageModel.select(message);
+                return index == 0;
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java]: IMO should be return true. !fix

File: app/src/main/java/com/teamagam/gimelgimel/app/model/entities/messages/InMemory/InMemoryMessagesModel.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import com.teamagam.gimelgimel.app.model.entities.messages.MessagesModel;
+
+import java.util.ArrayList;
+
+/**
+ * {@link MessagesModel} class that holds its messages in-memory
+ */
+public class InMemoryMessagesModel extends NotifyingDataChangedObservable implements MessagesModel {
+
+    private ArrayList<Message> mMessages = new ArrayList<>();
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/entities/messages/InMemory/InMemoryMessagesModel.java:14]: those model classes should be singeltons

Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/entities/messages/InMemory/InMemoryMessagesModel.java:14]: !resolve

File: app/src/test/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModelTest.java [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    public void getDisplayedMessageRandomAccessorWithNoSelectedMessage_shouldReturnFirstMessageAsSelected() throws Exception {
+        //Arrange
+        mMessagesModel.add(createMessage());
+        mMessagesModel.add(createMessage());
+
+        //Act
+        MessagesViewModel.DisplayedMessagesRandomAccessor res = mMessageViewModel.getDisplayedMessagesRandomAccessor();
+
+        //Assert
+        assertThat(res.get(0).isSelected(), is(true));
```
Comment (teamagam1) [inline app/src/test/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModelTest.java]: what should happen if I do the followin test:
        assertThat(res.get(1).isSelected(), is(true));
        assertThat(res.get(0).isSelected(), is(false));
I think this is OK

File: app/src/test/java/com/teamagam/gimelgimel/app/model/entities/messages/InMemoryMessagesModelTest.java [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+        Message m3 = createMockMessage();
+        int idx1 = mMessageModel.add(mMessage);
+        int idx2 = mMessageModel.add(m2);
+        int idx3 = mMessageModel.add(m3);
+
+        //Act
+        mMessageModel.remove(idx2);
+
+        //Assert
+        assertThat(mMessageModel.get(idx1), sameInstance(mMessage));
```
Comment (teamagam1) [inline app/src/test/java/com/teamagam/gimelgimel/app/model/entities/messages/InMemoryMessagesModelTest.java:104]: the indices shouldnt be consistent at all!

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+/**
+ * Messages view-model for messages presentation use-case
+ */
+public class MessagesViewModel extends NotifyingDataChangedObservable implements DataChangedObservable {
+
+    private MessagesModel mMessageModel;
+    private SelectedMessageModel mSelectedMessageModel;
+    private MessagesReadStatusModel mMessagesReadStatusModel;
+
+    public MessagesViewModel(
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java:21]: How the VM notifies the View when data is updated? !resolve

Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java:21]: How the VM notifies the View when data is updated? !resolve

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/DisplayMessage.java (hunk 6 above)
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/DisplayMessage.java:11]: OK i got it, never mind

File: app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+        mObserver = observer;
+    }
+
+    public void notifyObserver(){
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java]: change to protected !fix

File: app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java (hunk 4 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java]: !fix

File: app/src/main/java/com/teamagam/gimelgimel/app/common/DataChangedObservable.java (hunk 1 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/common/DataChangedObservable.java:1]: Its a common class.

File: app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java (hunk 4 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java]: Sababa Achi

Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java]: !fixed

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/Message.java (hunk 5 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/Message.java:46]: None of them are used, and that's the way it should be. We wouldn't want anyone to change the type/content, id, sender id or creation date  of an existing message.

File: app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java (hunk 14 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/common/NotifyingDataChangedObservable.java]: !fixed

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/Message.java (hunk 7 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/Message.java:70]: Its generated by the server.

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java (hunk 8 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java:15]: But its not a base viewmodel class.
" I dont like" is not a valid argument. !resolved

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java (hunk 13 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java:21]: !resolved

Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java:21]: Through the NotifyingDataChangedObservable

File: app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java (hunk 9 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModel.java]: It shouldn't. It should return true for the first message (or some other default message if you can think of anything more reasonable) but not all messages should be marked as read. !fixed

File: app/src/main/java/com/teamagam/gimelgimel/app/model/entities/messages/InMemory/InMemoryMessagesModel.java (hunk 10 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/model/entities/messages/InMemory/InMemoryMessagesModel.java:14]: They dont need to be singletons. We'll only create one instance of them. !resolved

File: app/src/test/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModelTest.java (hunk 11 above)
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessagesViewModelTest.java]: fixed that.

**Pull request merged**